google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc_test.cc

```diff
@@ -0,0 +1,758 @@
+#include <chrono>
+#include <deque>
+#include <iostream>
+#include <memory>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace pubsublite_internal {
+namespace {
+
+using ::google::cloud::testing_util::IsOk;
+using ::testing::_;
+using ::testing::StrictMock;
+
+using ::google::cloud::internal::RetryPolicy;
+
+struct FakeRequest {
+  std::string key;
+};
+
+struct FakeResponse {
+  std::string key;
+  std::string value;
+};
+
+// Used as part of the EXPECT_CALL() below:
+bool operator==(FakeRequest const& lhs, FakeRequest const& rhs) {
+  return lhs.key == rhs.key;
+}
+
+bool operator==(FakeResponse const& lhs, FakeResponse const& rhs) {
+  return lhs.key == rhs.key && lhs.value == rhs.value;
+}
+
+using MockAsyncStreamReturnType =
+std::unique_ptr<AsyncStreamingReadWriteRpc<FakeRequest, FakeResponse>>;
+
+class MockAsyncReaderWriter
+    : public AsyncStreamingReadWriteRpc<FakeRequest, FakeResponse> {
+ public:
+  MOCK_METHOD(future<absl::optional<FakeResponse>>, Read, (), (override));
+  MOCK_METHOD(future<bool>, Write, (FakeRequest const&, grpc::WriteOptions),
+              (override));
+  MOCK_METHOD(future<bool>, WritesDone, (), (override));
+  MOCK_METHOD(void, Cancel, (), (override));
+  MOCK_METHOD(future<Status>, Finish, (), (override));
+  MOCK_METHOD(future<bool>, Start, (), (override));
+};
+
+class MockRetryPolicy : public RetryPolicy {
+ public:
+  MOCK_METHOD(bool, OnFailure, (Status const&), (override));
+  MOCK_METHOD(bool, IsExhausted, (), (const, override));
+  MOCK_METHOD(bool, IsPermanentFailure, (Status const&), (const, override));
+};
+
+class MockStub {
+ public:
+  MOCK_METHOD(MockAsyncStreamReturnType, FakeStream, (), ());
+  MOCK_METHOD(std::unique_ptr<RetryPolicy>, FakeRetryPolicy, (), ());
+};
+
+std::unique_ptr<BackoffPolicy> DefaultBackoffPolicy() {
+  using us = std::chrono::microseconds;
+  return ExponentialBackoffPolicy(/*initial_delay=*/us(10),
+      /*maximum_delay=*/us(40), /*scaling=*/2.0)
+      .clone();
+}
+
+future<void> MockSleeper(std::chrono::duration<double>) {
+  return make_ready_future();
+}
+
+const Status kFailStatus = Status(StatusCode::kUnavailable, "Unavailable");
+
+const FakeResponse kBasicResponse = FakeResponse{"key0", "value0_0"};
+const FakeResponse kBasicResponse1 = FakeResponse{"key0", "value0_1"};
+
+const FakeRequest kBasicRequest = FakeRequest{"key0"};
+
+#define STANDARD_SINGLE_RETRY_POLICY \
+EXPECT_CALL(mock, FakeRetryPolicy) \
+      .WillOnce( \
+          []() { return absl::make_unique<StrictMock<MockRetryPolicy>>(); }) \
+      .WillOnce([]() { \
+        auto mock_retry_policy = \
+            absl::make_unique<StrictMock<MockRetryPolicy>>(); \
+        EXPECT_CALL(*mock_retry_policy, IsExhausted).WillOnce([]() { \
+          return false; \
+        }); \
+        EXPECT_CALL(*mock_retry_policy, OnFailure(_)).WillOnce([]() { \
+          return true; \
+        }); \
+        return mock_retry_policy; \
+      });
```

Comment:
No, never use macros if there's any other option. You can just make this a function